### PR TITLE
[Rust] Introduce new module definitions and APIs in wasmedge-sys

### PR DESCRIPTION
**Pull request by apepkuss**
### Motivation
This PR is the replacement of #846. The PR introduces some new module definitions and safe APIs in wasmedge-sys; in addition, it also brings about some improvements to the existing module definitions and the APIs. All the changes introduced reference the definitions in [wasmedge.h.](https://github.com/WasmEdge/WasmEdge/blob/master/include/api/wasmedge/wasmedge.h).

### Details
The following iterms are involved:
- Refactored items
  - Config
  - ImportObject
    - [x] #746 
    - [x] #747 
    - [x] #749 

  - Function
  - Table
    - [x] #797 

  - Memory
    - [x] #767 

  - Global
    - [x] #733 
    - [x] #736 
    - [x] #741 

  - Module
  - Vm
    - [x] #770 
    - [x] #771 
    - [x] #772 
    - [x] #779 

  - error.rs

- New Items
  - Executor
    - [x] #700 
    - [x] #701 
    - [x] #702 
    - [x] #703 
    - [x] #704 
    - [x] #705 
    - [x] #709

  - Statistics
    - [x] #790 
    - [x] #791 
    - [x] #792 
    - [x] #793 
    - [x] #794 
    - [x] #840 
    - [x] #841 

  - Store
    - [x] #796
    - [x] #813 
    - [x] #814
    - [x] #815
    - [x] #816 
    - [x] #817 
    - [x] #818 
    - [x] #819 
    - [x] #820 
    - [x] #821 
    - [x] #822 
    - [x] #823 
    - [x] #824 
    - [x] #825 
    - [x] #826 
    - [x] #827 
    - [x] #828 
    - [x] #829 
    - [x] #830 
    - [x] #831 
    - [x] #832 
    - [x] #833 
    - [x] #834 
    - [x] #835 
    - [x] #836 
    - [x] #837 
    - [x] #838 
    - [x] #839 

  - types.rs
    - MemType
      - [x] #786 
      - [x] #787
      - [x] #789  

  - utils.rs

### Other Issues
This PR also relates to the following issues:
- #601 
- #635 
- #628 
- #846

### CODE REVIEW

File: bindings/rust/wasmedge-sys/src/module.rs [hunk 1]
```diff
@@ -1,41 +1,45 @@
...
+        })
     }
 
-    pub fn load_from_buffer(
-        config: &crate::config::Config,
-        buffer: Vec<u8>,
-    ) -> Result<Self, ErrReport> {
+    pub fn load_from_buffer(config: &crate::config::Config, buffer: &[u8]) -> WasmEdgeResult<Self> {
         let loader_ctx = unsafe { wasmedge::WasmEdge_LoaderCreate(config.ctx) };
         let mut ctx: *mut wasmedge::WasmEdge_ASTModuleContext = std::ptr::null_mut();
```
Comment (ZhangHanDong) [inline bindings/rust/wasmedge-sys/src/module.rs:45]: nice! This assertion is guaranteed to be safety. 
This unsafe function should be marked with a document comment to indicate that it is safe, as is the usual practice. 
```rust
/// # Safety 
///  Avoid the problem of uninitialized memory due to the introduction of an empty buffer
```

File: bindings/rust/wasmedge-sys/src/module.rs [hunk 2]
```diff
@@ -1,41 +1,45 @@
...
+            ctx,
+            registered: false,
+        })
     }
 
-    pub fn load_from_buffer(
-        config: &crate::config::Config,
-        buffer: Vec<u8>,
-    ) -> Result<Self, ErrReport> {
+    pub fn load_from_buffer(config: &crate::config::Config, buffer: &[u8]) -> WasmEdgeResult<Self> {
```
Comment (ZhangHanDong) [inline bindings/rust/wasmedge-sys/src/module.rs:42]: nice! This assertion is guaranteed to be safety. 
This unsafe function should be marked with a document comment to indicate that it is safe, as is the usual practice. 
```rust
/// # Safety 
///  Avoid the problem of uninitialized memory due to the introduction of an empty buffer
```

File: bindings/rust/wasmedge-sys/src/import_obj.rs [hunk 3]
```diff
@@ -1,60 +1,129 @@
...
+        }
+        global.registered = true;
+        global.ctx = std::ptr::null_mut();
+    }
+
+    pub fn init_wasi<T: Iterator<Item = E>, E: AsRef<str>>(
+        &self,
+        args: Option<T>,
+        envs: Option<T>,
+        dirs: Option<T>,
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:71]: Thanks for your comment. According to the definition of WasmEdge_ImportObjectInitWASI in wasmedge.h, dirs is an optional argument, so we currently declare it as Option. If the spec would be revised to remove the argument, we'd remove it in the next revision.

[the remark this answers is not in the document]
Comment (hydai) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:71]: The `Dir` related arguments are removed from the argument list: https://github.com/WasmEdge/WasmEdge/blob/master/include/api/wasmedge/wasmedge.h#L2377-L2380

Ref: #843

File: bindings/rust/wasmedge-sys/src/store.rs [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+    string::StringRef,
+};
+
+#[derive(Debug)]
+pub struct Store {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_StoreContext,
+    pub(crate) registered: bool,
+}
+impl Store {
+    pub fn create() -> Option<Self> {
```
Comment (yanganto) [inline bindings/rust/wasmedge-sys/src/store.rs:13]: same here, thanks.

File: bindings/rust/wasmedge-sys/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+pub fn vec_string_to_c_array<T: Iterator<Item = E>, E: AsRef<str>>(
+    args: T,
+) -> *const *const std::os::raw::c_char {
+    let mut v = vec![];
+    for arg in args {
+        v.push(string_to_c_char(arg));
+    }
+    let p = v.as_ptr();
+    std::mem::forget(v);
```
Comment (yanganto) [inline bindings/rust/wasmedge-sys/src/utils.rs:30]: Because we are using `forget` here, it is good to add the comment on `fn vec_string_to_c_array` to make sure others will take care of the memory in manually way.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 6]
```diff
@@ -17,39 +15,67 @@ pub struct Vm {
 }
 
 impl Vm {
-    pub fn create(config: &crate::config::Config) -> Self {
-        let ctx = unsafe {
-            wasmedge::WasmEdge_VMCreate(config.ctx, std::ptr::null_mut() /* store */)
+    pub fn create(config: Option<&Config>, store: Option<&Store>) -> Option<Self> {
```
Comment (yanganto) [inline bindings/rust/wasmedge-sys/src/vm.rs:18]: same here

**Review by yanganto — CHANGES_REQUESTED** [review]

File: bindings/rust/wasmedge-sys/src/error.rs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct WasmEdgeError {
+    /// 0x00: Success
+    /// 0x01: Terminated -> Success
+    /// 0x02: Failed
+    /// 0x03: NullError
+    /// 0x20: File not found
+    pub code: u32,
+    pub message: String,
+}
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/error.rs:14]: Accepted. We'll refactor the existing WasmEdgeError. Thanks!

File: bindings/rust/wasmedge-sys/src/error.rs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        let code = unsafe { WasmEdge_ResultGetCode(result) };
+        let message = unsafe {
+            let c_str = std::ffi::CStr::from_ptr(WasmEdge_ResultGetMessage(result));
+            c_str.to_string_lossy().into_owned()
+        };
+        WasmEdgeError { code, message }
+    }
+}
+
+pub type WasmEdgeResult<T> = Result<T, WasmEdgeError>;
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/error.rs:34]: WasmEdgeResult is a short form of Result<T, WasmEdgeError>, so that we can use WasmEdgeResult<_concrete_type_> instead of WasmEdgeResult<_concrete_type_, Error>. That would make the return type more concise. Thanks!

File: bindings/rust/wasmedge-sys/src/import_obj.rs [hunk 9]
```diff
@@ -1,60 +1,110 @@
...
 #[derive(Debug)]
 pub struct ImportObj {
     pub(crate) ctx: *mut wasmedge::WasmEdge_ImportObjectContext,
 }
 
 impl ImportObj {
-    pub fn create(module_name: impl AsRef<str>) -> Self {
-        let raw_module_name: wasmedge::WasmEdge_String =
-            StringRef::from(module_name.as_ref()).into();
+    pub fn create(name: impl AsRef<str>) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:13]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/executor.rs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    string::StringRef,
+    Config, ImportObj, Module, Statistics, Store,
+};
+use std::ptr;
+
+pub struct Executor {
+    ctx: *mut wasmedge::WasmEdge_ExecutorContext,
+}
+impl Executor {
+    pub fn create(conf: Option<&Config>, stat: Option<&mut Statistics>) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/executor.rs:13]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/import_obj.rs [hunk 11]
```diff
@@ -1,60 +1,110 @@
...
 
-    pub fn add_func(&mut self, func_name: impl AsRef<str>, func: &mut Function) {
-        let raw_func_name: wasmedge::WasmEdge_String = StringRef::from(func_name.as_ref()).into();
+    pub fn add_func(&mut self, name: impl AsRef<str>, func: &mut Function) {
+        let raw_func_name: wasmedge::WasmEdge_String = StringRef::from(name.as_ref()).into();
         unsafe {
             wasmedge::WasmEdge_ImportObjectAddFunction(self.ctx, raw_func_name, (*func).ctx);
         }
-        func.registed = true;
+        func.registered = true;
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:27]: :)

File: bindings/rust/wasmedge-sys/src/instance/function.rs [hunk 12]
```diff
@@ -113,26 +113,26 @@ impl Function {
...
-        if !self.registed {
+        if !self.registered && !self.ctx.is_null() {
             unsafe { wasmedge::WasmEdge_FunctionInstanceDelete(self.ctx) };
         }
     }
 }
 
 #[derive(Debug)]
-pub struct Type {
+pub struct FuncType {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/function.rs:131]: I think the prefix is necessary, as there is also a memory type, when we use them to declare new variables, the type name with a prefix can improve the readability for developers; in addition, "Type" has a wider semantics, "FuncType" could be more concrete.

File: bindings/rust/wasmedge-sys/src/instance/global.rs [hunk 13]
```diff
@@ -1,11 +1,73 @@
...
+};
 
+#[derive(Debug)]
 pub struct Global {
-    ctx: *mut wasmedge::WasmEdge_GlobalInstanceContext,
+    pub(crate) ctx: *mut wasmedge::WasmEdge_GlobalInstanceContext,
+    pub(crate) registered: bool,
 }
+impl Global {
+    pub fn create(ty: &mut GlobalType, val: Value) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/global.rs:12]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/instance/global.rs [hunk 14]
```diff
@@ -1,11 +1,73 @@
...
     fn drop(&mut self) {
-        unsafe { wasmedge::WasmEdge_GlobalInstanceDelete(self.ctx) };
+        if !self.registered && !self.ctx.is_null() {
+            unsafe { wasmedge::WasmEdge_GlobalInstanceDelete(self.ctx) };
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct GlobalType {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/global.rs:46]: The explanation is the same as FuncType.

File: bindings/rust/wasmedge-sys/src/instance/global.rs [hunk 15]
```diff
@@ -1,11 +1,73 @@
...
+    pub(crate) ctx: *mut wasmedge::WasmEdge_GlobalTypeContext,
+    pub(crate) registered: bool,
+}
+impl GlobalType {
+    pub fn create(val_ty: ValType, mutable: Mutability) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/global.rs:51]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/instance/memory.rs [hunk 16]
```diff
@@ -1,11 +1,59 @@
...
+use crate::{types::Limit, wasmedge};
 
+#[derive(Debug)]
 pub struct Memory {
-    ctx: *mut wasmedge::WasmEdge_MemoryInstanceContext,
+    pub(crate) ctx: *mut wasmedge::WasmEdge_MemoryInstanceContext,
+    pub(crate) registered: bool,
+}
+impl Memory {
+    pub fn create(limit: Limit) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/memory.rs:9]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/instance/memory.rs [hunk 17]
```diff
@@ -1,11 +1,59 @@
...
     fn drop(&mut self) {
-        unsafe { wasmedge::WasmEdge_MemoryInstanceDelete(self.ctx) };
+        if !self.registered && !self.ctx.is_null() {
+            unsafe { wasmedge::WasmEdge_MemoryInstanceDelete(self.ctx) };
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct MemType {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/memory.rs:31]: The explanation is the same as FuncType

File: bindings/rust/wasmedge-sys/src/instance/memory.rs [hunk 18]
```diff
@@ -1,11 +1,59 @@
...
+    pub(crate) ctx: *mut wasmedge::WasmEdge_MemoryTypeContext,
+    pub(crate) registered: bool,
+}
+impl MemType {
+    pub fn create(limit: Limit) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/memory.rs:36]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/instance/table.rs [hunk 19]
```diff
@@ -1,11 +1,35 @@
...
+};
 
+#[derive(Debug)]
 pub struct Table {
-    ctx: *mut wasmedge::WasmEdge_TableInstanceContext,
+    pub(crate) ctx: *mut wasmedge::WasmEdge_TableInstanceContext,
+    pub(crate) registered: bool,
+}
+impl Table {
+    pub fn create(ref_type: WasmEdgeRefType, limit: Limit) -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/table.rs:12]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/statistics.rs [hunk 20]
```diff
@@ -0,0 +1,56 @@
+use super::wasmedge;
+
+pub struct Statistics {
+    pub(crate) ctx: *mut wasmedge::WasmEdge_StatisticsContext,
+    pub(crate) registered: bool,
+}
+impl Statistics {
+    pub fn create() -> Option<Self> {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/statistics.rs:8]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/vm.rs (hunk 6 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/vm.rs:18]: Accepted. Thanks!

File: bindings/rust/wasmedge-sys/src/import_obj.rs [hunk 21]
```diff
@@ -1,60 +1,110 @@
...
-            .map(|env| CString::new(env.as_ref()).unwrap())
-            .collect();
-
-        let cstr_preopens: Vec<_> = preopens
-            .iter()
-            .map(|preopen| CString::new(preopen.as_ref()).unwrap())
-            .collect();
-
-        let ctx = unsafe {
-            wasmedge::WasmEdge_ImportObjectCreateWASI(
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:48]: Accepted. I'll add this function. Thanks!

File: bindings/rust/wasmedge-sys/src/import_obj.rs [hunk 22]
```diff
@@ -63,3 +113,50 @@ impl Drop for ImportObj {
...
+mod tests {
+    use super::ImportObj;
+    use crate::{
+        instance::Function,
+        io::{I1, I2},
+        Value,
+    };
+
+    #[test]
+    fn test_imports_add_host_function() {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/import_obj.rs:127]: Removed the test case. Thanks!

File: bindings/rust/wasmedge-sys/src/module.rs [hunk 23]
```diff
@@ -1,41 +1,45 @@
...
 #[derive(Debug)]
 pub struct Module {
     pub(crate) ctx: *mut wasmedge::WasmEdge_ASTModuleContext,
+    pub(crate) registered: bool,
 }
-
 impl Drop for Module {
     fn drop(&mut self) {
-        unsafe { wasmedge::WasmEdge_ASTModuleDelete(self.ctx) };
+        if !self.registered && !self.ctx.is_null() {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/module.rs:16]: Accepted. The first condition will be removed; and I'll refactor the relevant code.

File: bindings/rust/wasmedge-sys/src/types.rs [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+    fn from(ty: WasmEdgeRefType) -> Self {
+        match ty {
+            WasmEdgeRefType::FuncRef => wasmedge::WasmEdge_RefType_FuncRef,
+            WasmEdgeRefType::ExternRef => wasmedge::WasmEdge_RefType_ExternRef,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Limit {
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/types.rs:23]: Accepted. Will implement two From traits: one for Range to WasmEdge_Limit, one for WasmEdge_limit to Range. Thanks!

File: bindings/rust/wasmedge-sys/src/utils.rs (hunk 5 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/utils.rs:30]: Will remove this function definition as it's not used at all. Thanks!

File: bindings/rust/wasmedge-sdk/examples/hello.rs [hunk 25]
```diff
@@ -2,33 +2,33 @@
```
Comment (yanganto) [inline bindings/rust/wasmedge-sdk/examples/hello.rs:2]: Because `wasmedge-sdk` crate will be dropped and may be redesigned in the future.  
This comment is acceptable in this PR.

File: bindings/rust/wasmedge-sdk/src/config.rs [hunk 26]
```diff
@@ -1,14 +1,27 @@
...
 pub struct Config {
     pub(crate) inner: wasmedge::Config,
 }
-
 impl Config {
-    pub fn with_wasi() -> Self {
-        let config = Self::default();
-        let inner: wasmedge::Config = config.inner.enable_wasi();
-        Self { inner }
+    pub fn create() -> Option<Self> {
```
Comment (yanganto) [inline bindings/rust/wasmedge-sdk/src/config.rs:8]: May we use `Result` for this return value?  Thanks.

File: bindings/rust/wasmedge-sdk/src/config.rs [hunk 27]
```diff
@@ -1,14 +1,27 @@
...
+        let inner = wasmedge::Config::create();
+        match inner {
+            Ok(inner) => Some(Self { inner }),
+            Err(_) => None,
+        }
+    }
+}
+impl Config {
+    pub fn with_wasi() -> Option<Self> {
```
Comment (yanganto) [inline bindings/rust/wasmedge-sdk/src/config.rs:17]: May we use Result for this return value? Thanks.

File: bindings/rust/wasmedge-sys/src/vm.rs (hunk 6 above)
Comment (yanganto) [reply bindings/rust/wasmedge-sys/src/vm.rs:18]: May we use `result` return type for this?
Thanks.

File: bindings/rust/wasmedge-sys/src/vm.rs [hunk 28]
```diff
@@ -166,11 +193,143 @@ impl Vm {
...
 
 impl Drop for Vm {
     fn drop(&mut self) {
-        unsafe { wasmedge::WasmEdge_VMDelete(self.ctx) };
-        self.import_objects.drain(..);
+        if !self.ctx.is_null() {
+            unsafe { wasmedge::WasmEdge_VMDelete(self.ctx) };
+        }
+        if !self.import_objects.is_empty() {
+            self.import_objects.drain(..);
```
Comment (yanganto) [inline bindings/rust/wasmedge-sys/src/vm.rs:212]: Is there any issue to drain an empty vector?
Or is there any beneficial to drain a vector before checking it is empty or not?

**Review by yanganto — APPROVED** [review]
We almost there, thanks for modification so quick.

File: bindings/rust/wasmedge-sys/src/vm.rs (hunk 28 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/vm.rs:212]: As for the drain method, it can work with an empty Vec instance. The import_objects field is introduced in the original code. To avoid destroying the original design logic, keep it there in this PR. It would be better to decide to either remove it or refactor it in the future. Thanks!

File: bindings/rust/wasmedge-sdk/examples/hello.rs (hunk 25 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/examples/hello.rs:2]: Thanks!

File: bindings/rust/wasmedge-sdk/src/config.rs (hunk 26 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/src/config.rs:8]: Accepted. Thanks!

File: bindings/rust/wasmedge-sdk/src/config.rs (hunk 27 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/src/config.rs:17]: Accepted. Thanks!

File: bindings/rust/wasmedge-sdk/examples/hello.rs (hunk 25 above)
Comment (hydai) [reply bindings/rust/wasmedge-sdk/examples/hello.rs:2]: We should add some comments/git commit message to explain the reason. It looks like just comment out the example without any reason in both commit history and comments.

File: bindings/rust/wasmedge-sdk/examples/quickstart.rs [hunk 29]
```diff
@@ -2,23 +2,23 @@
 
 #[cfg_attr(test, test)]
 fn main() -> Result<(), Box<dyn std::error::Error>> {
-    let module_path = std::path::PathBuf::from(env!("WASMEDGE_DIR"))
-        .join("tools/wasmedge/examples/fibonacci.wasm");
+    // let module_path = std::path::PathBuf::from(env!("WASMEDGE_DIR"))
```
Comment (hydai) [inline bindings/rust/wasmedge-sdk/examples/quickstart.rs:5]: Ditto.

File: bindings/rust/wasmedge-sdk/examples/hello.rs (hunk 25 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/examples/hello.rs:2]: Accepted. Will add a comment for it. Thanks!

File: bindings/rust/wasmedge-sdk/examples/quickstart.rs (hunk 29 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/examples/quickstart.rs:5]: Accepted. Will add a comment for it. Thanks!

File: bindings/rust/wasmedge-sdk/src/vm.rs [hunk 30]
```diff
@@ -76,10 +76,12 @@ impl<'a> VmBuilder<'a> {
 
     pub fn create(self) -> Result<Vm<'a>, anyhow::Error> {
         let vm = self.inner;
+        // let mut inner_module = vm.module.inner;
```
[the remark this answers is not in the document]
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/src/vm.rs:79]: Accepted. Will remove it, thanks!

File: bindings/rust/wasmedge-sys/src/instance/function.rs [hunk 31]
```diff
@@ -58,8 +58,8 @@ extern "C" fn wraper_fn(
 #[derive(Debug)]
 pub struct Function {
     pub(crate) ctx: *mut wasmedge::WasmEdge_FunctionInstanceContext,
-    pub(crate) registed: bool,
-    ty: Option<Type>,
+    pub(crate) registered: bool,
+    pub(crate) ty: Option<FuncType>,
```
Comment (hydai) [inline bindings/rust/wasmedge-sys/src/instance/function.rs:62]: And I would like to discuss with the naming of `Type`s. Do we have to add the prefix in each `Type` such as `FuncType`. Rust has namespace, would it be better to use `Func::Type` in this case?

File: bindings/rust/wasmedge-sys/src/vm.rs (hunk 28 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/vm.rs:212]: Will remove the redundant check code in line 211, thanks!

File: bindings/rust/wasmedge-sys/src/instance/function.rs (hunk 31 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/instance/function.rs:62]: For now, wasmedge-sys involves several types: FuncType in function, MemType in memory, and ValType in types. We surely can distinguish these three by their namespace. However, when these types are used in the same code file and multiple times, that could become a little bit verbose, like function::Type. Of course, aliasing can help. But, if the case happens in several code files, users have to do it in each code file. In addition, (personal idea), the semantics implied by function::Type is not equal to the one by FuncType.

File: bindings/rust/wasmedge-sys/src/store.rs (hunk 4 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sys/src/store.rs]: Accepted. Thanks!

File: bindings/rust/wasmedge-sdk/src/config.rs [hunk 32]
```diff
@@ -1,14 +1,28 @@
...
 pub struct Config {
     pub(crate) inner: wasmedge::Config,
 }
-
 impl Config {
-    pub fn with_wasi() -> Self {
-        let config = Self::default();
-        let inner: wasmedge::Config = config.inner.enable_wasi();
-        Self { inner }
+    pub fn create() -> Result<Self, ConfigError> {
```
Comment (yanganto) [inline bindings/rust/wasmedge-sdk/src/config.rs:9]: If a user not put any additional modification to new an instance, s/he may like to use the `default` of config by Rust convention.  

We use create because the name is from C-Api, it may be nice to provide `default` function of `Default` trait to alias the create without parameters, or we can do it in different PR.

**Review by yanganto — APPROVED** [review]

**Review by yanganto — DISMISSED** [review]

File: bindings/rust/wasmedge-sdk/src/config.rs (hunk 32 above)
Comment (apepkuss) [reply bindings/rust/wasmedge-sdk/src/config.rs:9]: Yes, Default is more convenient and friendly to Rust developers. We can consider introducing Default trait for such cases in future individual revisions.

**Comment by hydai** [comment]
Hmmm. The same issue.

Looks like your commit history contains others' commits. Please fix it. You can use push forcely to your origin branch.
截圖 2021-12-03 at 13 30 34@2x

**Review by hydai — CHANGES_REQUESTED** [review]
Your commit history should not contain others' commits. Please fix it.

**Comment by apepkuss** [comment]
> Your commit history should not contain others' commits. Please fix it.

@hydai I rebased the PR branch so it contains others' commits. I'll fix it.

**Review by yanganto — APPROVED** [review]

**Review by hydai — APPROVED** [review]
